File: backend/spring/src/main/java/nidhal/stackoverflowstatus/controller/StackExchangeApiController.java
package nidhal.stackoverflowstatus.controller;

import jakarta.annotation.PostConstruct;
import lombok.AllArgsConstructor;
import nidhal.stackoverflowstatus.configuration.CacheConfig;
import nidhal.stackoverflowstatus.configuration.StackExchangeApiConfig;
import nidhal.stackoverflowstatus.dao.QuestionDao;
import nidhal.stackoverflowstatus.service.StackExchangeApiService;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@RestController
@AllArgsConstructor
@RequestMapping("/api/v2/questions")
public class StackExchangeApiController {

    private final StackExchangeApiService stackExchangeApiService;
    private final StackExchangeApiConfig stackExchangeApiConfig;
    private final QuestionDao questionDao;
    private final CacheConfig cacheConfig;

    private List<String> programmingLanguages;

    @PostConstruct
    private void init() {
        programmingLanguages = stackExchangeApiConfig.getProgrammingLanguages();
    }

    private static final long ONE_DAY_AGO_IN_SECONDS = Instant.now().minus(Duration.ofDays(1)).getEpochSecond();
    private static final long ONE_WEEK_AGO_IN_SECONDS = Instant.now().minus(Duration.ofDays(7)).getEpochSecond();
    private static final int MAX_THREADS = 10;

    @GetMapping("/this-week")
    public String storeThisWeekQuestions() {

        // Since our application store only the questions from the last week, we need to delete the old ones
        questionDao.deleteAllByCreationDateLessThan(ONE_WEEK_AGO_IN_SECONDS);

        // clear the cache before updating the questions
        cacheConfig.clearCache();

        try (ExecutorService executorService = Executors.newFixedThreadPool(MAX_THREADS)) {
            // store all questions related to each programming language in a separate thread
            for (String language : programmingLanguages) {
                CompletableFuture.runAsync(() -> stackExchangeApiService.storeAllQuestionRelatedTo(language, ONE_WEEK_AGO_IN_SECONDS), executorService);
            }
        }

        return "This week's questions up to date";
    }


    @GetMapping("/this-day")
    public String storeThisDayQuestions() {

        // Since our application store only the questions from the last week, we need to delete the old ones
        questionDao.deleteAllByCreationDateLessThan(ONE_WEEK_AGO_IN_SECONDS);

        // clear the cache before updating the questions
        cacheConfig.clearCache();

        try (ExecutorService executorService = Executors.newFixedThreadPool(MAX_THREADS)) {
            // store all questions related to each programming language in a separate thread
            for (String language : programmingLanguages) {
                CompletableFuture.runAsync(() -> stackExchangeApiService.storeAllQuestionRelatedTo(language, ONE_DAY_AGO_IN_SECONDS), executorService);
            }
        }
        
        return "Today's questions up to date";
    }
}

File: backend/dotnet/stackoverflow_statistics/Controllers/StackExchangeApiController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Options;
using stackoverflow_statistics.Configuration;
using stackoverflow_statistics.Services;


namespace stackoverflow_statistics.Controllers
{
    [ApiController]
    [Route("api/v1/questions")]
    public class StackExchangeApiController : ControllerBase
    {
        private readonly StackExchangeApiService _stackExchangeApiService;
        private readonly List<string> _programmingLanguages;

        public StackExchangeApiController(
            StackExchangeApiService stackExchangeApiService,
            IOptions<StackExchangeApiConfig> stackExchangeApiConfig)
        {
            _stackExchangeApiService = stackExchangeApiService;
            _programmingLanguages = stackExchangeApiConfig.Value.ProgrammingLanguages;
        }

        private static readonly long OneDayAgoInSeconds = DateTimeOffset.UtcNow.AddDays(-1).ToUnixTimeSeconds();
        private static readonly long OneWeekAgoInSeconds = DateTimeOffset.UtcNow.AddDays(-7).ToUnixTimeSeconds();

        [HttpGet("this-week")]
        public async Task<IActionResult> StoreThisWeekQuestions()
        {
            foreach (var language in _programmingLanguages)
            {
                await _stackExchangeApiService.StoreAllQuestionsRelatedToAsync(language, OneWeekAgoInSeconds);
            }

            return Ok("This week's questions up to date");
        }

        [HttpGet("this-day")]
        public async Task<IActionResult> StoreThisDayQuestions()
        {
            foreach (var language in _programmingLanguages)
            {
                await _stackExchangeApiService.StoreAllQuestionsRelatedToAsync(language, OneDayAgoInSeconds);
            }

            return Ok("Today's questions up to date");
        }
    }
}